app/services/badges/award_contributor_from_github.rb

```diff
@@ -9,19 +14,42 @@ class AwardContributorFromGithub
       "forem/DEV-ios",
     ].freeze
 
-    def self.call(since = 1.day.ago, msg = "Thank you so much for your contributions!")
-      return unless (badge_id = Badge.id_for_slug(BADGE_SLUG))
+    def self.call(msg = "Thank you so much for your contributions!")
+      new(msg).call
+    end
 
-      REPOSITORIES.each do |repo|
-        commits = Github::OauthClient.new.commits(repo, since: since.utc.iso8601)
+    def initialize(msg)
+      @msg = msg
+    end
 
-        authors_uids = commits.map { |commit| commit.author.id }
-        Identity.github.where(uid: authors_uids).find_each do |i|
-          BadgeAchievement
-            .where(user_id: i.user_id, badge_id: badge_id)
-            .first_or_create(rewarding_context_message_markdown: msg)
+    def call
+      REPOSITORIES.each do |repo|
+        commits = Github::OauthClient.new.commits(repo)
```

Comment:
One efficiency concern - this used to return just the last 1 days worth of commits for each repo, now it returns a long list of all commits (which are in pages of 100, and we have over 9000 commits in the forem/forem repo).

Assuming the rate limiting is not an issue (it was for me, locally, since I had not setup github authentication tokens in my .env file, it looks like it was active for you as well based on the VCR cassette) there is still a question of space and time used by us locally.

```ruby
[6] pry(main)> Benchmark.measure { commits = Github::OauthClient.new.commits(repo, since: 1.day.ago) }
=> #<Benchmark::Tms:0x0000558876410970
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=0.3045605589868501,
 @stime=0.007760000000000003,
 @total=0.05495700000000005,
 @utime=0.047197000000000044>
```

This was to get a page of 7 commits from the last day - but it took about 300ms.

I tried to ask for the full forem/forem repo to confirm the behavior - and saw a series of paged requests, one by one, until I had hit the 60 requests per hour api rate limit for unauthenticated requests. I got 5700 out of 9546 commits back (the client handled the 403 response by returning the accumulated results, which were most recent first). 

I ran it again after the rate limit expired, making 60 requests took 49.7 seconds, and that's only 2/3rds of the time needed - so this job would require about 1m20s to pull the repositories data down.

We currently run this on a schedule every hour (in addition to any other github api requests we make elsewhere, like Readme preview rendering for linked  repositories) so there's a possible chance that as the number of commits or repos grows, the authenticated rate limit (5000 requests per hour) may be exceeded. Right now we would expect only 100-200 requests from this service when it runs (total number of commits from four repositories) but if we added public repositories to track for awards, or the number of commits continued to grow, this might present some limitations later.

It would be much nicer from an execution time and memory use standpoint if we could limit the request to a window (like the 1.day.ago filter that was here) and increment a repo:commits counter for the user when a new commit is seen, that probably adds some complexity, or use another endpoint (`stats/contributors`, Octokit exposes this as `contributor_stats(repo)`) 

https://docs.github.com/en/rest/reference/repos#get-all-contributor-commit-activity documents this, `total` and `author.id` look like they are the relevant items we want and are equivalent to fetching all commits, and counting by author id. 

It might be worthwhile to validate the counts are the same or comparable between the two endpoints, but this appears to be a single request with an aggregated response (or a smaller set of requests) already grouped by github user, if we're planning to compare cumulative commit counts rather than looking for new committers (as we had done previously).

Reply:
I tested the contributors stats, and I get a single page (of 100) -  this appears to be by design https://docs.github.com/en/github/visualizing-repository-data-with-graphs/accessing-basic-repository-data/viewing-a-projects-contributors

Right now this limits the view to contributors with four or more commits to the repo (in forem/forem) and is probably going to miss all first time contributors. Would it make sense to pass through this twice per repo, once with the last day's new commits, for first time contributors, and again using the statistics endpoint for cumulative badges?

Reply:
Doing a little digging - it looks like passing the anon=true option (`contributors("forem/forem", true)` in Octokit's client) gives 618 contributors rather than the last 100. 

There were 178 nil ids here, but this includes the long tail of users with a single commit, and only takes 7 requests (paginated by 100). 

I propose we rework the request to use this instead of commits(repo):

```ruby
Github::OauthClient.new.contributors("forem/forem", true) # anon=true for full contributor list instead of summary graph
```

The response format changes (so this is a slightly larger rework to the api response handling than I would have liked), we're dealing with an array of objects like this:

```
{:login=>"dependabot[bot]",
 :id=>49699333,
 :node_id=>"MDM6Qm90NDk2OTkzMzM=",
 :avatar_url=>"https://avatars.githubusercontent.com/in/29110?v=4",
 :gravatar_id=>"",
 :url=>"https://api.github.com/users/dependabot%5Bbot%5D",
 :html_url=>"https://github.com/apps/dependabot",
 :followers_url=>"https://api.github.com/users/dependabot%5Bbot%5D/followers",
 :following_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/following{/other_user}",
 :gists_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/gists{/gist_id}",
 :starred_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/starred{/owner}{/repo}",
 :subscriptions_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/subscriptions",
 :organizations_url=>"https://api.github.com/users/dependabot%5Bbot%5D/orgs",
 :repos_url=>"https://api.github.com/users/dependabot%5Bbot%5D/repos",
 :events_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/events{/privacy}",
 :received_events_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/received_events",
 :type=>"Bot",
 :site_admin=>false,
 :contributions=>1093}
```

I assume `id` and `contributions` are the key factors needed for the result here, the results are in descending order of contributions, and the sum of the returned contributions is equal to the number of commits shown on the repo page in github so it included everything.

Reply:
sounds good, let me take a look at this

Reply:
@djuber Made changes as requested.